### PR TITLE
Initial specs migration.

### LINKED ISSUE #792: Migrate specs over to app

### PR DESCRIPTION
**Pull request by adlerjohn**
Fixes #792

An initial migration of the existing specs as of [`e59efd63a2165866584833e91e1cb8a6ed8c8203`](https://github.com/celestiaorg/celestia-specs/tree/e59efd63a2165866584833e91e1cb8a6ed8c8203). It's expected that there are a few broken links and extra content that has to be pruned, along with incorrect content. The important thing is to not miss anything.

Also sets up a CI workflow that will build the book.

### CODE REVIEW

File: specs/src/specs/data_structures.md
Comment (rootulp) [inline specs/src/specs/data_structures.md:86]: [no change needed] I'm not sure why height is a signed integer rather than unsigned but this conforms to tendermint/tendermint ([ref](https://github.com/tendermint/tendermint/blob/aa303edaefeb8a0856d7494cca24e51f6f3e2393/state/errors.go#L13)) so no change needed

File: specs/theme/highlight.js
Comment (rootulp) [inline specs/theme/highlight.js:2]: [no change needed] this file looks :suspect: but it seems intentional
1. https://rust-lang.github.io/mdBook/format/theme/syntax-highlighting.html
2. https://highlightjs.org/
3. https://github.com/celestiaorg/celestia-specs/blob/master/theme/highlight.js

File: specs/README.md
Comment (rootulp) [inline specs/README.md:5]: [optional]

```suggestion
Prerequisite: [install mdBook](https://rust-lang.github.io/mdBook/guide/installation.html)

To build book:
```

**Review by rootulp — DISMISSED** [review]
LGTM! Thanks for your hard work on this

**Review by evan-forbes — APPROVED** [review]
dope, glad these are finding a home.

File: specs/README.md
Comment (evan-forbes) [reply specs/README.md:5]: this is a good addition, but I'd prefer not to add it in this PR

File: specs/src/specs/consensus.md
Comment (evan-forbes) [reply specs/src/specs/consensus.md:72]: same as https://github.com/celestiaorg/celestia-app/pull/793#discussion_r980619637

File: specs/src/specs/proto/wire.proto
Comment (liamsi) [inline specs/src/specs/proto/wire.proto:86]: This Tx type does not exist. Neither any "graffiti" field. 
I am not sure that copying the spec as is without going through the hard work to update it is the best approach but up to you and the core/app team.

**Review by liamsi — COMMENTED** [review]
I would ask to add John, Matt, and me to the `specs` sub-dir as code-owner.

File: specs/src/specs/proto/types.proto
Comment (liamsi) [inline specs/src/specs/proto/types.proto:264]: Can be deleted I assume?

**Review by liamsi — COMMENTED** [review]
> Also sets up a CI workflow that will push the built book to a main directory on the gh-pages branch on each merge to main, allowing a rendered form of the book to be displayed on GitHub pages.

I would recommend to remove this until the specs are somewhat up-to-date again. Otherwise, we will be confusing a bunch of folks visiting the published specs again.

**Comment by musalbas** [comment]
Shouldn't the parts about how data is encoded be in Celestia core, not app? We want to have specs be in the repo that they're relevant for.

**Comment by musalbas** [comment]
Also, it seems a large part of the spec is outdated (sparse Merkle trees, graffiti, etc). IMO there's not much advantage to merging the existing outdated specs into app, as people can already reference the existing specs repo. I think there should be a draft PR to do a basic update and clean up of the specs, as well as decoupling the bits for core and app, so that we have a general structure that can be built on, before merging this.

**Comment by evan-forbes** [comment]
> I think there should be a draft PR to do a basic update and clean up of the specs, as well as decoupling the bits for core and app, so that we have a general structure that can be built on, before merging this.

@musalbas I think there was a miscommunication, where those of us on the call agreed to move the specs over first, and then edit them with individual PRs. this is why Rootul and myself approved so quickly and why John made the PR in the first place.

if we don't want to do that, then we can turn this PR into a draft and make the individual PRs on this branch until we're comfortable with merging.

> Shouldn't the parts about how data is encoded be in Celestia core, not app?

all encoding code is kept in the application now, so I think it makes sense to keep these specs here.

> I would recommend to remove this until the specs are somewhat up-to-date again. Otherwise, we will be confusing a bunch of folks visiting the published specs again.

not publishing the specs until they match the implementation makes a lot of sense. We can split this portion into a different PR.

**Review by evan-forbes — CHANGES_REQUESTED** [review]
we should split the CI publishing mechanism into a different PR https://github.com/celestiaorg/celestia-app/pull/793#pullrequestreview-1121612747

**Comment by musalbas** [comment]
Makes sense, would it make sense to merge this into a different branch like `specs` in that case, until it's ready for `main`?

File: specs/src/specs/block_proposer.md
Comment (adlerjohn) [reply specs/src/specs/block_proposer.md:26]: We do!

File: specs/src/specs/data_structures.md
Comment (adlerjohn) [reply specs/src/specs/data_structures.md:86]: (Invalid) reasoning here: https://blog.cosmos.network/choosing-a-type-for-blockchain-height-beware-of-unsigned-integers-714804dddf1d

File: specs/theme/highlight.js
Comment (adlerjohn) [reply specs/theme/highlight.js:2]: It's copy-pasted from the original repo without change.

**Comment by adlerjohn** [comment]
1. Made a new branch `specs-staging` that we can squash-merge PRs into, starting with this one, and re-targeted this PR on that branch. This avoids end-users seeing a WIP `specs` directory.
2. Removed the deploy step from CI, added a build step to ensure the book actually builds properly.

**Review by evan-forbes — APPROVED** [review]

**Comment by evan-forbes** [comment]
Thanks @adlerjohn! gets my :+1: :+1:

File: specs/README.md
Comment (adlerjohn) [reply specs/README.md:5]: Fixed in https://github.com/celestiaorg/celestia-app/pull/800

**Pull request merged**
